### PR TITLE
Create a new C API KernelContext_GetAllocator() for Custom Op scenario

### PR DESCRIPTION
**Pull request by jslhcl**
### Description
Create a new C API KernelContext_GetAllocator() for Custom Op scenario

### Motivation and Context
Create a new C API KernelContext_GetAllocator() for Custom Op scenario

### CODE REVIEW

File: onnxruntime/core/session/custom_ops.cc [hunk 1]
```diff
@@ -106,6 +106,17 @@ ORT_API_STATUS_IMPL(OrtApis::KernelContext_GetGPUComputeStream, _In_ const OrtKe
   API_IMPL_END
 };
 
+ORT_API_STATUS_IMPL(OrtApis::KernelContext_GetAllocator, _In_ const OrtKernelContext* context, _In_ const OrtMemoryInfo* mem_info, _Outptr_ OrtAllocator** out) {
+  API_IMPL_BEGIN
+  onnxruntime::AllocatorPtr allocator = reinterpret_cast<const onnxruntime::OpKernelContextInternal*>(context)->GetAllocator(*mem_info);
+  if (!allocator) {
+    return OrtApis::CreateStatus(ORT_INVALID_ARGUMENT, "No requested allocator available");
+  }
+  *out = new onnxruntime::OrtAllocatorImplWrappingIAllocator(std::move(allocator));
```
Comment (RandySheriffH) [inline onnxruntime/core/session/custom_ops.cc:115]: >new [](http://example.com/codeflow?start=9&length=3)

raw "new" is going to trigger warnings in static analysis. usually rely on unique_ptr for creations, and use .release() to acquire the raw thereafter.

File: include/onnxruntime/core/session/onnxruntime_c_api.h [hunk 2]
```diff
@@ -4150,6 +4150,18 @@ struct OrtApi {
...
+   *
+   * \param[in] context OrtKernelContext instance
+   * \param[in] mem_info OrtMemoryInfo instance
+   * \param[out] out A pointer to OrtAllocator.
+   *
+   * \snippet{doc} snippets.dox OrtStatus Return Value. Always returns nullptr.
+   *
+   * \since Version 1.15.
+   */
+  ORT_API2_STATUS(KernelContext_GetAllocator, _In_ const OrtKernelContext* context, _In_ const OrtMemoryInfo* mem_info, _Outptr_ OrtAllocator** out);
```
Comment (RandySheriffH) [inline include/onnxruntime/core/session/onnxruntime_c_api.h:4164]: >KernelContext_GetAllocator [](http://example.com/codeflow?start=18&length=26)

we have a wrapper in cxx api for kernel context:
[https://github.com/microsoft/onnxruntime/blob/1f7c2f724ffa6362f4d7f3f4360531458bd70d82/include/onnxruntime/core/session/onnxruntime_cxx_api.h#L1674](https://github.com/microsoft/onnxruntime/blob/1f7c2f724ffa6362f4d7f3f4360531458bd70d82/include/onnxruntime/core/session/onnxruntime_cxx_api.h#L1674)
might be good to extend corresponding rountines there for easier consumption.

File: onnxruntime/core/framework/op_kernel_context_internal.h [hunk 3]
```diff
@@ -71,6 +71,10 @@ class OpKernelContextInternal : public OpKernelContext {
 
   const bool& GetTerminateFlag() const noexcept { return terminate_flag_; }
 
+  AllocatorPtr GetAllocator(const OrtMemoryInfo& memory_info) const {
```
[the remark this answers is not in the document]
Comment (jslhcl) [reply onnxruntime/core/framework/op_kernel_context_internal.h:74]: For now, it is difficult. The new API is defined in the subclass OpKernelContextInternal, while the two functions you mentioned are in base class OpKernelContext, which doesn't have session_state_ member variable. So these two functions are getting allocators from execution_frame_

However, there are two subclasses (ExcutionFrame, OptimizedExecutionFrame) derived from IExecutionFrame, and OptimizedExecutionFrame get the allocator from EP, not from session_state.

I think we can add a TODO item for this work once the allocator refactor work is done (thus both subclasses of IExecutionFrame get allocators from session_state. And we also need to define this new API in the base class OpKernelContext.

File: onnxruntime/core/session/custom_ops.cc (hunk 1 above)
Comment (jslhcl) [reply onnxruntime/core/session/custom_ops.cc:115]: It follows the implementation of CreateAllocator in allocator_adapters.cc. Shall we follow the convention by adding 

#if defined(_MSC_VER) && !defined(__clang__)
#pragma warning(disable : 26409)
#endif

to suppress the warning?

**Review by souptc — APPROVED** [review]

**Pull request merged**